maven-core/src/main/java/org/apache/maven/plugin/internal/DefaultPluginDependenciesResolver.java

```diff
@@ -300,4 +311,40 @@ public boolean visitLeave(DependencyNode node) {
             return true;
         }
     }
+
+    private static class CoreDependencySelector implements DependencySelector {
+        private final HashSet<String> coreArtifacts;
+
+        private CoreDependencySelector(final Collection<String> coreArtifacts) {
+            this.coreArtifacts = new HashSet<>(coreArtifacts);
+        }
+
+        @Override
+        public boolean selectDependency(final org.eclipse.aether.graph.Dependency dependency) {
```

Comment:
I'm not a big fan of the `final` keyword on arguments and local variable... They are actually only useful on fields.